### PR TITLE
Operator Forwarder Production Readiness

### PR DESCRIPTION
**Pull request by austinborn**
- Add new tests to test suite
- Update interface names `AuthorizedReceiverInterface` -> `IAuthorizedReceiver` and `WithdrawalInterface` -> `IWithdrawal`
- Move all files out of `dev/` folder
- Update compiler files and gas snapshot

### CODE REVIEW

File: contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  bytes4 private constant ORACLE_REQUEST_SELECTOR = Operator.oracleRequest.selector;
+  bytes4 private constant OPERATOR_REQUEST_SELECTOR = Operator.operatorRequest.selector;
+  uint256 private constant SELECTOR_LENGTH = 4;
+  uint256 private constant EXPECTED_REQUEST_WORDS = 2;
+  uint256 private constant MINIMUM_REQUEST_LENGTH = SELECTOR_LENGTH + (32 * EXPECTED_REQUEST_WORDS);
+  bytes receivedData;
+
+  address _spoofedLinkToken = address(0xCCBBAA);
+
+  function test_FuzzOnTokenTransferFromLink(
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol:17]: This entire test feels convoluted and I'm not sure what it's supposed to do. It looks like you're testing the test, not the contract code?

File: contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      mstore(add(data, 68), amount) // ensure correct amount is passed0.8.19
+    }
+    // solhint-disable-next-line avoid-low-level-calls
+    (bool success, ) = address(this).delegatecall(data); // calls oracleRequest
+    require(success, "Unable to create request");
+  }
+
+  // @dev Reverts if the given data does not begin with the `oracleRequest` function selector
+  // @param data The data payload of the request
+  modifier permittedFunctionsForLINK(bytes memory data) {
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol:72]: Why is this a modifier?

File: contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      // solhint-disable-next-line avoid-low-level-calls
+      funcSelector := mload(add(data, 32))
+    }
+    _validateTokenTransferAction(funcSelector, data);
+    _;
+  }
+
+  // @notice Require that the token transfer action is valid
+  // @dev OPERATOR_REQUEST_SELECTOR = multiword, ORACLE_REQUEST_SELECTOR = singleword
+  function _validateTokenTransferAction(bytes4 funcSelector, bytes memory data) internal pure {
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol:84]: Why isn't this function inlined?

File: contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    require(data.length >= MINIMUM_REQUEST_LENGTH, "Invalid request length");
+    require(
+      funcSelector == OPERATOR_REQUEST_SELECTOR || funcSelector == ORACLE_REQUEST_SELECTOR,
+      "Must use whitelisted functions"
+    );
+  }
+
+  // Callback function for oracle request fulfillment
+  fallback(bytes calldata _receivedData) external returns (bytes memory) {
+    receivedData = _receivedData;
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol:94]: Why are we setting global state in a test when we can just return information from these helper functions?

File: contracts/src/v0.8/operatorforwarder/test/operator.t.sol [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+import {Operator} from "../Operator.sol";
+import {ChainlinkClientHelper} from "./testhelpers/ChainlinkClientHelper.sol";
+import {LinkToken} from "../../shared/token/ERC677/LinkToken.sol";
+
+import "./testhelpers/Deployer.t.sol";
+import "../AuthorizedReceiver.sol";
+
+contract OperatorTest is Deployer, AuthorizedReceiver {
+  address public s_link;
+  uint256 private dataReceived;
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/operator.t.sol:14]: Nit: private vars also need to start with s_

File: contracts/src/v0.8/operatorforwarder/test/operator.t.sol [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    address[] memory auth = new address[](1);
+    auth[0] = address(this);
+    s_operator = new Operator(s_link, address(this));
+    s_operator.setAuthorizedSenders(auth);
+
+    dataReceived = 0;
+  }
+
+  // Callback function for oracle request fulfillment
+  function callback(bytes32 _requestId) public {
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/operator.t.sol:32]: I prefer using a helper contract over using the test files themselves as receivers for callbacks

File: contracts/src/v0.8/operatorforwarder/test/operator.t.sol [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+  }
+
+  // @notice concrete implementation of AuthorizedReceiver
+  // @return bool of whether sender is authorized
+  function _canSetAuthorizedSenders() internal view override returns (bool) {
+    return true;
+  }
+
+  function test_Success(uint96 payment) public {
+    payment = uint96(bound(payment, 1, type(uint96).max));
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/operator.t.sol:44]: Nit: Usually bound is significantly preferred over assume, in this case I would have used assume as there is only a single value you exclude here, which is 0

File: contracts/src/v0.8/operatorforwarder/test/operator.t.sol [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+    bytes32 specId = keccak256("testSpecForFulfill");
+    bytes4 callbackFunctionId = bytes4(keccak256("callback(bytes32)"));
+    uint256 nonce = 1;
+    uint256 dataVersion = 1;
+    bytes memory dataBytes = "";
+    uint256 payment = 1 ether;
+    uint256 expiration = block.timestamp + 5 minutes;
+
+    // Convert bytes to bytes32
+    bytes32 data = bytes32(uint256(keccak256(dataBytes)));
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/operator.t.sol:181]: No need to convert to unit and then bytes32 again

File: contracts/src/v0.8/operatorforwarder/test/operator.t.sol [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    uint256 nonce = 2;
+    uint256 dataVersion = 1;
+    bytes memory dataBytes = "";
+    uint256 payment = 1 ether;
+    uint256 expiration = block.timestamp + 5 minutes;
+
+    uint256 withdrawableBefore = s_operator.withdrawable();
+
+    // Convert bytes to bytes32
+    bytes32 data = bytes32(uint256(keccak256(dataBytes)));
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/operator.t.sol:226]: Same

File: contracts/src/v0.8/operatorforwarder/test/operator.t.sol [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+
+    // No withdrawable balance as it's all locked
+    assertEq(s_operator.withdrawable(), withdrawableBefore, "Internal accounting not updated correctly");
+
+    bytes32 requestId = keccak256(abi.encodePacked(bob, nonce));
+
+    vm.startPrank(alice);
+    vm.expectRevert(bytes("Params do not match request ID"));
+    s_operator.cancelOracleRequest(requestId, payment, callbackFunctionId, expiration);
+    vm.stopPrank();
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/operator.t.sol:255]: You don't have to stop pranks, you can simply start a new one

File: contracts/src/v0.8/operatorforwarder/test/testhelpers/Deployer.t.sol [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+import {AuthorizedForwarder} from "../../AuthorizedForwarder.sol";
+import {Operator} from "../../Operator.sol";
+import {OperatorFactory} from "../../OperatorFactory.sol";
+
+import {MockReceiver} from "./MockReceiver.sol";
+
+import {MockLinkToken} from "../../../mocks/MockLinkToken.sol";
+
+abstract contract Deployer is Test {
+  OperatorFactory public factory;
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/testhelpers/Deployer.t.sol:15]: I don't think these have to be public, and they should be prefixed with s_

File: contracts/src/v0.8/operatorforwarder/test/testhelpers/Deployer.t.sol [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  Operator public operator;
+  AuthorizedForwarder public forwarder;
+
+  MockLinkToken public link;
+  MockReceiver public mockReceiver;
+
+  address public owner = address(uint160(uint256(keccak256("owner"))));
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/testhelpers/Deployer.t.sol:22]: Prefer makeAddr

**Comment by alecgard** [comment]
Are the OperatorForwarder Foundry tests currently being run on PR?

We may need to add to https://github.com/smartcontractkit/chainlink/blob/develop/.github/workflows/solidity-foundry.yml#L34

File: contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol (hunk 1 above)
Comment (austinborn) [reply contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol:17]: Yeah I should've read this before adding it in. Removing now, no idea what it does.

File: contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol (hunk 2 above)
Comment (austinborn) [reply contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol:72]: Removing test

File: contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol (hunk 3 above)
Comment (austinborn) [reply contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol:84]: Removing test

File: contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol (hunk 4 above)
Comment (austinborn) [reply contracts/src/v0.8/operatorforwarder/test/FuzzOperator.t.sol:94]: Removing test

File: contracts/src/v0.8/operatorforwarder/test/operator.t.sol (hunk 5 above)
Comment (austinborn) [reply contracts/src/v0.8/operatorforwarder/test/operator.t.sol:14]: Good catch, thank you

**Comment by austinborn** [comment]
> I don't think the OperatorForwarder foundry tests are currently being run on PR - we may need to add to https://github.com/smartcontractkit/chainlink/blob/develop/.github/workflows/solidity-foundry.yml#L34

Thanks for the note, adding in now

File: contracts/src/v0.8/operatorforwarder/test/Factory.t.sol [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+contract FactoryTest is Deployer {
+  function setUp() public {
+    _setUp();
+
+    vm.startPrank(s_alice);
+  }
+
+  /**
+   * @dev Test the deployment of a new operator.
+   */
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/Factory.t.sol]: nit: prefer `///` over `/** **/`

File: contracts/src/v0.8/operatorforwarder/test/Factory.t.sol [hunk 14]
```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import "forge-std/Test.sol";
+
+import "./testhelpers/Deployer.t.sol";
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/Factory.t.sol]: nit: it's OK for tests but I'd still prefer explicit imports everywhere. Not blocking but imo a good practice

File: contracts/src/v0.8/operatorforwarder/test/Forwarder.t.sol [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+    s_forwarder.ownerForward(address(s_link), new bytes(0));
+
+    vm.prank(s_alice);
+    s_forwarder.ownerForward(address(s_link), abi.encodeWithSignature("balanceOf(address)", address(0)));
+  }
+
+  /**
+   * @dev Tests the behavior of transfer and accept ownership of the contract.
+   */
+  function test_TransferOwnershipWithMessage() public {
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/Forwarder.t.sol]: I prefer the test naming 

`test_(Fuzz_)?<function_being_tested>_Success`
`test_(Fuzz_)?<function_being_tested>_<error_name>_Revert`

File: contracts/src/v0.8/operatorforwarder/test/testhelpers/Deployer.t.sol [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+import {Operator} from "../../Operator.sol";
+import {OperatorFactory} from "../../OperatorFactory.sol";
+import {LinkToken} from "../../../shared/token/ERC677/LinkToken.sol";
+
+abstract contract Deployer is Test {
+  OperatorFactory internal s_factory;
+  LinkToken internal s_link;
+  MockReceiver internal s_mockReceiver;
+
+  address private s_owner = makeAddr("owner");
```
Comment (RensR) [inline contracts/src/v0.8/operatorforwarder/test/testhelpers/Deployer.t.sol]: I kinda prefer a constant set of owner, alice, bob. That way there's no `s_` and it's all caps. Makes it clearer imo, and doesn't allow to you accidentally overwrite them

File: contracts/src/v0.8/operatorforwarder/test/Factory.t.sol (hunk 13 above)
Comment (austinborn) [reply contracts/src/v0.8/operatorforwarder/test/Factory.t.sol]: I'm just going to remove these comments since the test name says the same thing

**Review by RensR — APPROVED** [review]

**Review by alecgard — APPROVED** [review]

**Pull request merged**